### PR TITLE
Change name to Steppers-> Optimisers and move them to a separate section. Use warning/info/error messages.

### PR DESCRIPTION
**Pull request by gcarleo**
This pull request mainly moves `Stepper` into a dedicate section, to be logically consistent with other input sections. 

It also introduces info/warning/error messages utilities, to be used in place of existing output messages. The goal is to have more compact expressions, without explicitly calling MPI primitives in all the classes where messages are needed.

### CODE REVIEW

**Comment by twesterhout** [comment]
I have two things to note:
1) There are a lot of formatting changes in this PR. This makes it difficult to extract the actual changes to the code and makes one wonder how the code should actually look like. I'd like to bring up the idea of adding a `.clang-format` file the authors are happy with and require running `clang-format` before pushing.
2) I noticed that quite a lot of calls to `InfoMessage` look more or less like this
```cpp
InfoMessage("<variable_name> = " + std::to_string(variable_name));
```
One can simplify it to
```cpp
InfoMessage() << "<variable_name> = " << variable_name << '\n';
```
by making `InfoMessage` return a reference to the output stream. I'm slightly concerned that in the initial approach when messages get trickier we'll end up reimplementing formatting, which `std::ostream` already has. Also, strictly speaking, the latter version should be faster as it avoids unnecessary memory allocations.

Other than that, I'm quite okay with the proposed changes.

**Comment by femtobit** [comment]
Since we are currently moving things: The name `Stepper` is very generic. I wonder whether there is a more descriptive name for the steppers that we could use to distinguish them, e.g., from time steppers for the ODE, and to make it clearer what they are for.

I'm however not sure what a better name would be. They do perform a single optimization step for the parameters using the gradient but `OptimizationStepper` sounds a bit strange to me.

**Comment by gcarleo** [comment]
@twesterhout indeed I had to reformat some of the existing code using clang-tidy (and the Google format, as specified in the `CONTRIBUTING.md` 

It would be nice to add an automatic reformatting of the code when pushing, but I am not sure there is a clean way of doing this (I didn't look too much into this issue though). But yes, I agree that it can be quite distracting if most of the changes are due to formatting...! 

Concerning the other point, I like your suggestion too. 

Do you think it is technically possible to do something like: 

```c++
InfoMessage << "<variable_name> = " << variable_name << '\n';
````

i.e. without the () when calling InfoMessage, and using it as some stream object? That would be a bit more elegant than using the constructor.

File: NetKet/Utils/messages.hpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+#ifndef NETKET_MESSAGES_HPP
+#define NETKET_MESSAGES_HPP
+
+#include <mpi.h>
+#include <string>
+
+namespace netket {
+
+void InfoMessage(const std::string & message) {
```
[the remark this answers is not in the document]
Comment (gcarleo) [reply NetKet/Utils/messages.hpp]: Yes, it makes sense, I will add one, once we converge on the format of those Messages

**Comment by gcarleo** [comment]
@femtobit , good point concerning `Stepper`'s name. We could also call them `Optimizer` (as done for example in Keras or similar software), but I don't like too much that solution either, because `Stepper`s are only one of the building blocks of the full optimizer. `OptimizerStepper` is good too, but it's maybe too specific?

**Comment by femtobit** [comment]
@gcarleo From the suggestions so far I think `OptimizerStepper` is the best as it is similar to what other libraries use for similar purposes, as you mention. From what I understand, I'd say it is not too specific because performing the steps for gradient-based optimization methods is pretty much what they are supposed to do, right?

**Comment by femtobit** [comment]
Regarding the output streams: Being able to write `InfoMessage << "..."` should be possible via a global stream object, but I think using call syntax (be it a function or constructor call) is quite common (e.g., [Qt](https://doc.qt.io/qt-5/qdebug.html#basic-use)) and possibly more flexible if we want to change something in the future.

The simplest approach probably would be something along the lines of
```cpp
// Null stream taken from the Stackoverflow answer linked above
class NullBuffer : public std::streambuf
{
public:
  int overflow(int c) { return c; }
};

class NullStream : public std::ostream
{
public:
    NullStream() : std::ostream(&m_sb)
    {}
private:
    NullBuffer m_sb;
};

// global null stream
NullStream nullstream;

std::ostream& InfoMessage()
{
    // get MPI rank
    if(rank == 0) {
        return std::cout;
    } else {
        return nullstream;
    }
}
```

**Comment by twesterhout** [comment]
@gcarleo Is it really that important to save typing two characters? It is certainly possible to avoid a function call and have a global reference. In this case, however, care needs to be taken to avoid ODR violations and proper initialisation.
@femtobit your approach with a null stream is, I think, the correct one. However, I'm afraid, your global `nullstream` won't do (ODR). Also, I'm not entirely happy with overriding `overflow` as now the stream will be in a `!good() == true` state for no reason. Seeing as we're not maintaining compatibility with Windows, one might be better off just opening `/dev/null` and writing to it instead.

**Comment by gcarleo** [comment]
@twesterhout , I said that would be more elegant, not necessarily "better" in terms of typing performance! 

I am not not too fond of the `/dev/null` solution, but I haven't quite grasped what's wrong with @femtobit 's solution

**Comment by twesterhout** [comment]
@gcarleo Alright, I'll leave it at your discretion then :)

As for the `NullStream`, `nullstream` variable should become a local static variable in `InfoMessage`. Alternatively, see [3.2 here](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/n4381.html) for another "safe" way of defining global references.
My second concern with this solution was the fact that `InfoMessage().good()` will almost always return false (for null streams). This will probably go unnoticed unless someone accidentally turns on exceptions for this stream... It still doesn't feel quite right to me.
Perhaps there is another way to discard characters, but I'm no expert in iostreams (and have no intention of becoming one :) ).

**Comment by gcarleo** [comment]
OK, I implemented the stream objects. I think that this solution is good. @femtobit , I am still thinking about the naming of `Stepper`. I kind of don't like `OptimizerStepper` because **all** the other names are short :)

**Comment by femtobit** [comment]
What do you think about `GradientStepper`, since the steppers all perform gradient-based optimization? (It's not really shorter but subjectively I prefer the sound to `OptimizerStepper` or similar and it is more specific.)

**Comment by gcarleo** [comment]
@femtobit  I am actually having second thoughts about `Optimizer`. After all I think it is not a bad choice. Indeed in the future it would even "force" refactoring a little bit the code to move the `Sr` and make it a stand-alone optimiser, which actually makes sense.

**Pull request merged**